Fix some issues with folded AArch64 features

JamieCunliffe opened the pull request:
In #91608 the `fp` feature was removed for AArch64 and folded into the `neon` feature, however disabling the `neon` feature doesn't actually disable the `fp` feature. If my understanding on that thread is correct it should do.

While doing this, I also noticed that disabling some features would disable features that it shouldn't. For instance enabling `sve` will enable `neon`, however, when disabling `sve` it would then also disable `neon`, I wouldn't expect disabling `sve` to also disable `neon`.

cc @workingjubilee

JamieCunliffe committed:
Tie neon with fp-armv8.

In #91608 the fp-armv8 feature was removed as it's tied to the neon
feature. However disabling neon didn't actually disable the use of
floating point registers and instructions, for this `-fp-armv8` is
required.

JamieCunliffe committed:
Only disable folded features when it makes sense.

Some features that are tied together only make sense to be folded
together when enabling the feature. For example on AArch64 sve and
neon are tied together, however it doesn't make sense to disable neon
when disabling sve.

oli-obk commented:
r? @Amanieu since you reviewed the other PR, too

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4006274565/jobs/6877606870) [(plain)](https://github.com/rust-lang/rust/commit/2caab238755df4e4faf9d65029361ddd63c4dc8b/checks/10877182069/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:ac593985615ec2ede58e132d2e21d2b1cbd6127c)
Download action repository 'rust-lang/simpleinfra@master' (SHA:7873766cb4a6a0bb00c54496556b38db3fffa5d6)
Complete job name: PR (x86_64-gnu-llvm-13, false, ubuntu-20.04-xl)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-llvm-13
---
i.....i..............i....i..ii.................iii........iii.i........i............... 88/403
....ii.................i............i..i...............i.....i..iiii.......i..i.....i..i 176/403
iiiiii.........i.iii....i..i......................i...ii...i.....ii..i..ii...i.......... 264/403
.....ii........................i.i.ii.i..i...........i..i....i....i..iii...........i...i 352/403
Some tests failed in compiletest suite=codegen mode=codegen host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
i..FF..FF.............iiiiiiii.i...................

---- [codegen] tests/codegen/tied-features-strength.rs#DISABLE_NEON stdout ----


error in revision `DISABLE_NEON`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-13/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.DISABLE_NEON/tied-features-strength.ll" "/checkout/tests/codegen/tied-features-strength.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC,DISABLE_NEON" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen/tied-features-strength.rs:13:18: error: DISABLE_NEON: expected string not found in input
// DISABLE_NEON: attributes #0 = { {{.*}} "target-features"="+outline-atomics,-neon,-fp-armv8,+v8a" }
                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.DISABLE_NEON/tied-features-strength.ll:1:1: note: scanning from here
; ModuleID = 'tied_features_strength.94bf27d7-cgu.0'
^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.DISABLE_NEON/tied-features-strength.ll:13:105: note: possible intended match here
attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "target-cpu"="generic" "target-features"="+outline-atomics,-neon,-fp-armv8" }


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.DISABLE_NEON/tied-features-strength.ll
Check file: /checkout/tests/codegen/tied-features-strength.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'tied_features_strength.94bf27d7-cgu.0' 
check:13'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: source_filename = "tied_features_strength.94bf27d7-cgu.0" 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            3: target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128" 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            4: target triple = "aarch64-unknown-linux-gnu" 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            5:  
check:13'0     ~
            6: ; tied_features_strength::test 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            8: define void @_ZN22tied_features_strength4test17hb9a86fea28e79362E() unnamed_addr #0 { 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            9: start: 
check:13'0     ~~~~~~~
           10:  ret void 
check:13'0     ~~~~~~~~~~
           11: } 
check:13'0     ~~
           12:  
check:13'0     ~
           13: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "target-cpu"="generic" "target-features"="+outline-atomics,-neon,-fp-armv8" } 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:13'1                                                                                                             ?                                                                              possible intended match
           14:  
check:13'0     ~
           15: !llvm.module.flags = !{!0, !1} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16:  
check:13'0     ~
           17: !0 = !{i32 7, !"PIC Level", i32 2} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           18: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:13'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



---- [codegen] tests/codegen/tied-features-strength.rs#ENABLE_SVE stdout ----

error in revision `ENABLE_SVE`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-13/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.ENABLE_SVE/tied-features-strength.ll" "/checkout/tests/codegen/tied-features-strength.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC,ENABLE_SVE" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen/tied-features-strength.rs:7:16: error: ENABLE_SVE: expected string not found in input
// ENABLE_SVE: attributes #0 = { {{.*}} "target-features"="+outline-atomics,+sve,+neon,+v8a" }
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.ENABLE_SVE/tied-features-strength.ll:1:1: note: scanning from here
; ModuleID = 'tied_features_strength.94bf27d7-cgu.0'
^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.ENABLE_SVE/tied-features-strength.ll:13:105: note: possible intended match here
attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "target-cpu"="generic" "target-features"="+outline-atomics,+sve,+neon" }


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.ENABLE_SVE/tied-features-strength.ll
Check file: /checkout/tests/codegen/tied-features-strength.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
           1: ; ModuleID = 'tied_features_strength.94bf27d7-cgu.0' 
check:7'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           2: source_filename = "tied_features_strength.94bf27d7-cgu.0" 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           3: target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128" 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           4: target triple = "aarch64-unknown-linux-gnu" 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           5:  
check:7'0     ~
           6: ; tied_features_strength::test 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           7: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           8: define void @_ZN22tied_features_strength4test17hb9a86fea28e79362E() unnamed_addr #0 { 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           9: start: 
check:7'0     ~~~~~~~
          10:  ret void 
check:7'0     ~~~~~~~~~~
          11: } 
check:7'0     ~~
          12:  
check:7'0     ~
          13: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "target-cpu"="generic" "target-features"="+outline-atomics,+sve,+neon" } 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:7'1                                                                                                             ?                                                                         possible intended match
          14:  
check:7'0     ~
          15: !llvm.module.flags = !{!0, !1} 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          16:  
check:7'0     ~
          17: !0 = !{i32 7, !"PIC Level", i32 2} 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          18: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:7'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



---- [codegen] tests/codegen/tied-features-strength.rs#DISABLE_SVE stdout ----

error in revision `DISABLE_SVE`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-13/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.DISABLE_SVE/tied-features-strength.ll" "/checkout/tests/codegen/tied-features-strength.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC,DISABLE_SVE" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen/tied-features-strength.rs:10:17: error: DISABLE_SVE: expected string not found in input
// DISABLE_SVE: attributes #0 = { {{.*}} "target-features"="+outline-atomics,-sve,+v8a" }
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.DISABLE_SVE/tied-features-strength.ll:1:1: note: scanning from here
; ModuleID = 'tied_features_strength.94bf27d7-cgu.0'
^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.DISABLE_SVE/tied-features-strength.ll:13:105: note: possible intended match here
attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "target-cpu"="generic" "target-features"="+outline-atomics,-sve" }


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.DISABLE_SVE/tied-features-strength.ll
Check file: /checkout/tests/codegen/tied-features-strength.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'tied_features_strength.94bf27d7-cgu.0' 
check:10'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: source_filename = "tied_features_strength.94bf27d7-cgu.0" 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            3: target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128" 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            4: target triple = "aarch64-unknown-linux-gnu" 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            5:  
check:10'0     ~
            6: ; tied_features_strength::test 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            8: define void @_ZN22tied_features_strength4test17hb9a86fea28e79362E() unnamed_addr #0 { 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            9: start: 
check:10'0     ~~~~~~~
           10:  ret void 
check:10'0     ~~~~~~~~~~
           11: } 
check:10'0     ~~
           12:  
check:10'0     ~
           13: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "target-cpu"="generic" "target-features"="+outline-atomics,-sve" } 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:10'1                                                                                                             ?                                                                   possible intended match
           14:  
check:10'0     ~
           15: !llvm.module.flags = !{!0, !1} 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16:  
check:10'0     ~
           17: !0 = !{i32 7, !"PIC Level", i32 2} 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           18: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



---- [codegen] tests/codegen/tied-features-strength.rs#ENABLE_NEON stdout ----

error in revision `ENABLE_NEON`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-13/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.ENABLE_NEON/tied-features-strength.ll" "/checkout/tests/codegen/tied-features-strength.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC,ENABLE_NEON" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen/tied-features-strength.rs:16:17: error: ENABLE_NEON: expected string not found in input
// ENABLE_NEON: attributes #0 = { {{.*}} "target-features"="+outline-atomics,+neon,+fp-armv8,+v8a" }
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.ENABLE_NEON/tied-features-strength.ll:1:1: note: scanning from here
; ModuleID = 'tied_features_strength.94bf27d7-cgu.0'
^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.ENABLE_NEON/tied-features-strength.ll:13:105: note: possible intended match here
attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "target-cpu"="generic" "target-features"="+outline-atomics,+neon,+fp-armv8" }


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/tied-features-strength.ENABLE_NEON/tied-features-strength.ll
Check file: /checkout/tests/codegen/tied-features-strength.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'tied_features_strength.94bf27d7-cgu.0' 
check:16'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: source_filename = "tied_features_strength.94bf27d7-cgu.0" 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            3: target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128" 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            4: target triple = "aarch64-unknown-linux-gnu" 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            5:  
check:16'0     ~
            6: ; tied_features_strength::test 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            8: define void @_ZN22tied_features_strength4test17hb9a86fea28e79362E() unnamed_addr #0 { 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            9: start: 
check:16'0     ~~~~~~~
           10:  ret void 
check:16'0     ~~~~~~~~~~
           11: } 
check:16'0     ~~
           12:  
check:16'0     ~
           13: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn "target-cpu"="generic" "target-features"="+outline-atomics,+neon,+fp-armv8" } 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:16'1                                                                                                             ?                                                                              possible intended match
           14:  
check:16'0     ~
           15: !llvm.module.flags = !{!0, !1} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16:  
check:16'0     ~
           17: !0 = !{i32 7, !"PIC Level", i32 2} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           18: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------




failures:
    [codegen] tests/codegen/tied-features-strength.rs#DISABLE_NEON
    [codegen] tests/codegen/tied-features-strength.rs#DISABLE_SVE
    [codegen] tests/codegen/tied-features-strength.rs#ENABLE_NEON
    [codegen] tests/codegen/tied-features-strength.rs#ENABLE_SVE
test result: FAILED. 321 passed; 4 failed; 78 ignored; 0 measured; 0 filtered out; finished in 4.32s

Build completed unsuccessfully in 0:13:03
```

</details>

JamieCunliffe committed:
Make v8a match optional in the test feature list.

Amanieu commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -149,6 +149,9 @@ pub fn time_trace_profiler_finish(file_name: &Path) {
+//
+// Note: The first feature in the list that is returned is the mapping to the feature that is
+// provided from the `s` parameter.
 pub fn to_llvm_features<'a>(sess: &Session, s: &'a str) -> SmallVec<[&'a str; 2]> {
```
I think it's a mistake to make this function do too much. Could the logic for implied/tied features be moved to a separate function and handled separately?

JamieCunliffe replied:
Yeah, I agree with that and was thinking the same while making the change. This was already handling the implied features before this change, so I just added `neon` and `fp-armv8` to it to fix that issue. The fact it's used in multiple places is why I created a different function for if they should be disabled together.

It was done this way to fix those bugs, however, long term I think this should be a graph structure to correctly show the relationships between the different features.

Amanieu replied:
Perhaps it would be clearer to make this function return a struct instead:
```rust
struct LlvmFeatures<'a> {
    /// Features that should be enabled/disabled.
    features: smallvec![&'a str],

    /// Features that are implicitly enabled when the feature is enabled.
    dependencies: smallvec![&'a str],
}
```

apiraino commented:
If I read correctly [the last comment](https://github.com/rust-lang/rust/pull/107294#discussion_r1096335693), this can be switched to waiting on author to incorporate changes. Feel free to request a review with `@rustbot ready`, thanks!

@rustbot author

Dylan-DPC commented:
@JamieCunliffe any updates on this?

JamieCunliffe committed:
Create a structure to define the features from to_llvm_features.

Rather than returning an array of features from to_llvm_features, return a structure that contains
the dependencies. This also contains metadata on how the features depend on each other to allow for
the correct enabling and disabling.

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 04efd0da1cfb438c5cb41b6d3637e719416faf0f has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 04efd0da1cfb438c5cb41b6d3637e719416faf0f -->
<!-- homu: {"type":"Approved","sha":"04efd0da1cfb438c5cb41b6d3637e719416faf0f","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

JamieCunliffe commented:
@Amanieu I'm not sure if it needs to be approved again, but I updated it to resolve the merge conflicts.

Amanieu commented:
@bors r+

rust-timer commented:
Finished benchmarking commit (52dd1cde59ee86e5d61469ee1befe09c132c777d): [comparison URL](https://perf.rust-lang.org/compare.html?start=b08148f6a76010ea3d4e91d61245aa7aac59e4b4&end=52dd1cde59ee86e5d61469ee1befe09c132c777d&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b08148f6a76010ea3d4e91d61245aa7aac59e4b4&end=52dd1cde59ee86e5d61469ee1befe09c132c777d&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b08148f6a76010ea3d4e91d61245aa7aac59e4b4&end=52dd1cde59ee86e5d61469ee1befe09c132c777d&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.1%  | [1.1%, 1.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -4.8% | [-4.8%, -4.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.1%, -1.1%] | 1     |
| All ❌✅ (primary)                 | -4.8% | [-4.8%, -4.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b08148f6a76010ea3d4e91d61245aa7aac59e4b4&end=52dd1cde59ee86e5d61469ee1befe09c132c777d&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [1.1%, 1.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.5% | [2.5%, 2.5%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [1.1%, 1.1%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=b08148f6a76010ea3d4e91d61245aa7aac59e4b4&end=52dd1cde59ee86e5d61469ee1befe09c132c777d&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 646.753s -> 646.927s (0.03%)
